skimage/segmentation/slic_superpixels.py

```diff
@@ -1,17 +1,119 @@
-from collections.abc import Iterable
+import warnings
+import functools
+import collections as coll
 import numpy as np
 from scipy import ndimage as ndi
+from scipy.spatial.distance import pdist, squareform
 
+from ._slic import (_slic_cython, _enforce_label_connectivity_cython)
 from ..util import img_as_float, regular_grid
-from ..segmentation._slic import (_slic_cython,
-                                  _enforce_label_connectivity_cython)
 from ..color import rgb2lab
 
 
+def _get_mask_centroids(mask, n_centroids, spacing=None):
+    """Find regularly spaced centroids on a mask.
+
+    Parameters
+    ----------
+    mask : 3D ndarray
+        The mask where the centroids must be positioned.
+    n_centroids : int
+        The number of centroids to be returned.
+    spacing : sequence of same, optional,
+        Spacing of elements along each dimension.
+
+    Returns
+    -------
+    centroids : 2D ndarray
+        The coordinates of the centroids with shape (n_centroids, 3).
+    steps : 1D ndarray
+        The approximate distance between two seeds in all dimensions.
+
+    """
+
+    # Get tight ROI around the mask to optimize
+    coord = np.asarray(np.nonzero(mask))
+    bbox = coord.min(-1), coord.max(-1) + 1
+    roi = mask[bbox[0][0]: bbox[1][0],
+               bbox[0][1]: bbox[1][1],
+               bbox[0][2]: bbox[1][2]].copy()
+
+    if spacing is None:
```

Comment:
Why not calculate default spacing and then use same distance metric for both?

Reply:
Sorry, I am not sure to understand what you mean here. `spacing` is an input to `slic` function.

Reply:
I mean, you can calculate `spacing` if it is not defined, and then use the same code path, whether it was specified or not.

Reply:
The use of `distance_transform_cdt` instead of `distance_transform_edt` was [suggested](https://github.com/scikit-image/scikit-image/pull/3850#discussion_r289618606) by @emmanuelle to reduce the computation time.
`spacing` is in the original API of `slic` and it is usable only with `distance_transform_edt`, that why I assumed that if the user doesn't specify `spacing`, the fast path must be used (`distance_transform_cdt`) and otherwise the path without approximation is used (`distance_transform_edt`).